rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -98,17 +100,77 @@ def __str__(self) -> str:
+        """The core method. Should be implemented by subclasses.
+        This method takes a new acquisition and decides where to insert it
+        and thus determines the PnL order.
+        """
+        ...
+
+    def processing_iterator(self) -> Iterator[AssetAcquisitionEvent]:
+        """Iteration method over acquisition events.
+        We can't return here Tuple of AssetAcquisitionEvents as we need to return
+        the first event each time but _acquisitions may be not modified between iterations.
+        """
+        while len(self._acquisitions) > 0:
+            yield self._acquisitions[0]
+
+    def get_acquisitions(self) -> Tuple[AssetAcquisitionEvent, ...]:
+        """Returns read-only _acquisitions"""
+        return tuple(self._acquisitions)
+
+    def consume_result(self, used_amount: FVal) -> None:
+        """This function should be used to consume results of the
+        currently processed event (received from __next__)
+        The current event's remaining_amount will be decreased by used_amount
+        If event's remaining_amount will become ZERO, the event will be deleted
+        May raise:
+        - IndexError if the method was called when acquisitions were empty
+        """
+        assert ZERO <= used_amount <= self._acquisitions[0].remaining_amount, \
+            f'Used amount must be in the interval [0, {self._acquisitions[0].remaining_amount}] but it was {used_amount}'  # noqa: E501p
+
+        self._acquisitions[0].remaining_amount -= used_amount
+        if self._acquisitions[0].remaining_amount == ZERO:
+            self._acquisitions.popleft()
+
+
+class FIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in FIFO (first-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        self._acquisitions.append(acquisition)
+
+
+class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in LIFO (last-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        self._acquisitions.appendleft(acquisition)
+
+
+class CostBasisEvents:
+    used_acquisitions: List[AssetAcquisitionEvent]
+    acquisitions_manager: BaseAcquisitionsOrder
+    spends: List[AssetSpendEvent]
+
+    def __init__(self, cost_basis_method: CostBasisMethod) -> None:
+        """This class contains data about acquisitions, spends and acquisitions
+        that were already used during accounting. `acquisitions` field contains a custom
+        Iterable that provides acquisitions in a particular order based on `cost_basis_method`"""
+        if cost_basis_method == CostBasisMethod.FIFO:
+            self.acquisitions_manager = FIFOAcquisitionsOrder()
+        elif cost_basis_method == CostBasisMethod.LIFO:
+            self.acquisitions_manager = LIFOAcquisitionsOrder()
+        else:
+            raise AssertionError(f'Unknown pnl order: {cost_basis_method}')
```

Comment:
Remove. This else can't happen. We don't need to have an else for enums. So long as all enum values are checked,